lib/logstash/outputs/elasticsearch.rb

```diff
@@ -413,11 +415,17 @@ def submit(actions)
       @submit_mutex.unlock
     end
     if bulk_response["errors"]
-      failed_actions = actions.select.with_index {|_,i| [429, 503].include?(bulk_response['statuses'][i]) }
-      unless failed_actions.empty?
-        @logger.debug "#{failed_actions.size}/#{actions.size} events were unsuccessful in sending"
-        retry_push(failed_actions)
+      actions_with_responses = actions.zip(bulk_response['statuses'])
+      actions_to_retry = []
+      actions_with_responses.each do |action, resp_code|
```

Comment:
:disappointed: there was no access the the exact response code previously. so the error wouldn't be able to let people know what exactly was the reason something went wrong.

Full retries are still managed at a higher level by StudBuffer. mostly by client issues relating to connection errors. I just named them to be more explicit that 429 and 502 are a part of the class of errors we decided are retryable.